### PR TITLE
Avoid unneeded copies when inlining controls

### PR DESCRIPTION
**Pull request by ChrisDodd**
When inlining controls, we check for aliasing and avoid creating copies of arguments if there is no possibility of aliasing, but currently that is restricted to a single call site.  This change extends that to multiple call sites if all the call sites use the same arguments.  If the arguments are different, we would need copies in at least some cases (to get them into the same place), but if all call sites are the same, copies are not needed.

Currently this only avoids the copies if the call sites are completely the same -- if the call sites had the same arguments for some parameters and not others, copies would be needed only for the parameters that were different between calls sites, but that would be a bit more work.

There's also a bunch of cleanups that crept into this branch.

### CODE REVIEW

File: frontends/p4/commonInlining.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ class AbstractInliner : public Transform {
 };
 
 template <class InlineList, class InlineWorkList>
-class InlineDriver : public Transform {
+class InlineDriver : public Visitor {
```
Comment (mbudiu-vmw) [inline frontends/p4/commonInlining.h]: This suggests that the driver does not change anything, which is not really true.

File: frontends/p4/inlining.cpp [hunk 2]
```diff
@@ -500,9 +500,20 @@ const IR::Node* GeneralInliner::preorder(IR::P4Control* caller) {
             // Use temporaries for these parameters
             std::set<const IR::Parameter*> useTemporary;
 
-            auto call = workToDo->uniqueCaller(inst);
+            const IR::MethodCallStatement *call = nullptr;
+            for (auto m : workToDo->callToInstance) {
+                if (m.second != inst) continue;
+                if (call) {
+                    if (!call->equiv(*m.first)) {
```
Comment (mbudiu-vmw) [inline frontends/p4/inlining.cpp]: is this precise enough to distinguish between two identifiers with the same name but different scopes?

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 3]
```diff
@@ -39,8 +39,7 @@ class ConstantTypeSubstitution : public Transform {
                              ReferenceMap* refMap,
                              TypeMap* typeMap) : subst(subst), refMap(refMap), typeMap(typeMap) {
         CHECK_NULL(subst); CHECK_NULL(refMap); CHECK_NULL(typeMap);
-        LOG3("ConstantTypeSubstitution " << subst);
-        setName("ConstantTypeSubstitution"); }
```
Comment (mbudiu-vmw) [inline frontends/p4/typeChecking/typeChecker.cpp]: I use --top4 passSubstring often; I hope that removing the setName won't affect that.

File: testdata/p4_14_samples_outputs/issue-1426-midend.p4 [hunk 4]
```diff
@@ -109,78 +108,20 @@ control ingress(inout headers hdr, inout metadata meta, inout standard_metadata_
         size = 1024;
         default_action = NoAction_7();
     }
-    @hidden action act() {
```
Comment (mbudiu-vmw) [inline testdata/p4_14_samples_outputs/issue-1426-midend.p4]: these are nice savings

File: frontends/p4/inlining.cpp (hunk 2 above)
Comment (ChrisDodd) [reply frontends/p4/inlining.cpp]: Can't happen as the previous UniqueNames pass will have renamed them to be different.

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 3 above)
Comment (ChrisDodd) [reply frontends/p4/typeChecking/typeChecker.cpp]: It fixes that -- when there are Visitor classes that inherit from each other, we want to use the name of the most-derived class, not the name of the base class.

File: frontends/p4/commonInlining.h (hunk 1 above)
Comment (ChrisDodd) [reply frontends/p4/commonInlining.h]: Visitor is the base class of all visitors, so encapsulates both visitors that change things and visitors that do not.  Deriving directly from Visitor means it does not use the normal Inspector/Modifier/Transform depth-first traversal.

Comment (ChrisDodd) [reply frontends/p4/commonInlining.h]: All of our Visitors *except* Inspectors (potentially) mutate the program.  Other direct Visitor subclasses like PassManager and PassRepeated may modify the program.  I guess we could add a comment to all such passes that they might modify the program, but it seems obvious from them not being Inspectors.

**Pull request merged**